File: examples/java/basic/src/main/java/com/rpl/agent/basic/TraceAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentTopology;
import com.rpl.agentorama.AgentsModule;
import com.rpl.agentorama.ops.RamaVoidFunction2;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;

/**
 * Java example demonstrating agent traces.
 *
 * <p>Features demonstrated:
 *
 * <ul>
 *   <li>agent-invoke: Synchronously invoke agent
 *   <li>Single-node agent execution
 *   <li>Basic agent tracing
 * </ul>
 *
 * <p>Uses String for request and response data.
 */
public class TraceAgent {

  /** Basic agent module with single node. */
  public static class TraceAgentModule extends AgentsModule {

    @Override
    protected void defineAgents(AgentTopology topology) {
      topology.newAgent("TraceAgent").node("process", null, new ProcessFunction());
    }
  }

  /** Function that processes input and returns result. */
  public static class ProcessFunction implements RamaVoidFunction2<AgentNode, String> {

    @Override
    public void invoke(AgentNode agentNode, String userName) {
      // Create a welcome message for the user
      String result = "Welcome to agent-o-rama, " + userName + "!";
      // Return the final result
      agentNode.result(result);
    }
  }

  public static void main(String[] args) throws Exception {
    try (InProcessCluster ipc = InProcessCluster.create()) {
      TraceAgentModule module = new TraceAgentModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("TraceAgent");

      System.out.println("Trace Agent Example:");

      // Invoke agent synchronously with sample user names
      String result1 = (String) agent.invoke("Alice");
      System.out.println("Result 1: " + result1);

      String result2 = (String) agent.invoke("Bob");
      System.out.println("Result 2: " + result2);

      String result3 = (String) agent.invoke("Charlie");
      System.out.println("Result 3: " + result3);

      System.out.println("\nNotice how:");
      System.out.println("- Agent processes each invocation");
      System.out.println("- Results are returned synchronously");
      System.out.println("- Agent traces can be viewed in the UI");
    }
  }
}

File: examples/clj/src/com/rpl/agent/basic/trace_agent.clj
(ns com.rpl.agent.basic.trace-agent
  "Demonstrates agent traces.

  Features demonstrated:
  - agent-invoke: Synchronously invoke agent"
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.agent-o-rama.impl.pobjects :as po]
   [com.rpl.agent-o-rama.impl.queries :as queries]
   [com.rpl.rama :as rama]
   [com.rpl.rama.path :as path]
   [com.rpl.rama.test :as rtest]))

;;; Basic agent module with single node
(aor/defagentmodule TraceAgentModule
  [topology]

  ;; Create agent with single node that processes input and returns result
  (-> topology
      (aor/new-agent "TraceAgent")
      (aor/node
       "process"
       nil
       (fn [agent-node user-name] ; user-name is the value in the agent-invoke
         ;; Create a welcome message for the user
         (let [result (str "Welcome to agent-o-rama, " user-name "!")]
           ;; Return the final result
           (aor/result! agent-node result))))))

(defn -main
  "Run the basic agent example with sample input"
  [& _args]
  ;; Create in-process cluster
  (with-open [ipc (rtest/create-ipc)]
    ;; launch the agent module
    (rtest/launch-module! ipc TraceAgentModule {:tasks 1 :threads 1})

    ;; Get agent manager and client
    (let [manager       (aor/agent-manager
                         ipc
                         (rama/get-module-name TraceAgentModule))
          agent         (aor/agent-client manager "TraceAgent")
          {:keys [task-id agent-invoke-id] :as init}
          (aor/agent-initiate agent "Alice")

          tracing-query
          (rama/foreign-query
           ipc
           (rama/get-module-name TraceAgentModule)
           (queries/tracing-query-name "TraceAgent"))
          root-pstate
          (rama/foreign-pstate
           ipc
           (rama/get-module-name TraceAgentModule)
           (po/agent-root-task-global-name "TraceAgent"))
          root-id
          (rama/foreign-select-one
           [(path/keypath agent-invoke-id) :root-invoke-id]
           root-pstate
           {:pkey task-id})
          result
          (rama/foreign-invoke-query
           tracing-query
           task-id
           [[task-id root-id]]
           1000)]

      ;; Invoke agent synchronously with sample user names
      (prn result))))

(comment

  (def ipc (rtest/create-ipc))
  (def ui (aor/start-ui ipc {:port 1975}))

  (rtest/launch-module! ipc TraceAgentModule {:tasks 1 :threads 1})
  (def manager
    (aor/agent-manager
     ipc
     (rama/get-module-name TraceAgentModule)))
  (def agent (aor/agent-client manager "TraceAgent"))

  (def init (aor/agent-initiate agent "Alice"))

  (def task-id (:task-id init))
  (def agent-invoke-id (:agent-invoke-id init))

  (def agent-result (aor/agent-result agent init))

  (def tracing-query
    (rama/foreign-query
     ipc
     (rama/get-module-name TraceAgentModule)
     (queries/tracing-query-name "TraceAgent")))

  (def root-pstate
    (rama/foreign-pstate
     ipc
     (rama/get-module-name TraceAgentModule)
     (po/agent-root-task-global-name "TraceAgent")))

  (def root-id
    (rama/foreign-select-one
     [(path/keypath agent-invoke-id) :root-invoke-id]
     root-pstate
     {:pkey task-id}))

  (def result
    (rama/foreign-invoke-query
     tracing-query
     task-id
     [[task-id root-id]]
     1000))
)
